### PR TITLE
builtin: change IError `msg` and `code` to methods + fix vlib

### PR DESCRIPTION
**Pull request by timbasel**
## Changes

This PR changes the `IError` interface from 
```v
interface IError {
  msg string
  code int
}
```
to
```v
interface IError {
  msg() string
  code() int
}
```

It also includes 2 default implementations for the interface: 
- `Error` provides empty methods returning the respective zero values (mostly used to avoid having to implement the `code()` method for each custom error)
- `MessageError` that works exactly like the old implementation and is returned by the `error()` function

Furthermore the PR fixes some 'compiler magic' that used the `msg` field to now call the `msg()` method, fixes vlib to use the `msg()` method and updates custom errors to use the improved syntax.

## Justifcation

This PR allows for more flexibility in how a custom error generates it message. For errors with fixed messages it also ensures consistency between the generated messages.

It also reduces the boilerplate required by a basic custom error:
```v
// allows for a custom error message at each call site
struct CustomError {
      MessageError
}

// enforces consistent error message for each error instance
struct FixedError {
      Error
}

fn (err FixedError) msg() string {
     return 'fixed error message'
}
```

A more complex example of this can also be seen in the changes to `vlib/flag/flag.v`
```v
struct UnknownFlagError {
	Error
	flag string
}

fn (err UnknownFlagError) msg() string {
	return 'Unknown flag `$err.flag`'
}

struct ArgsCountError {
	Error
	got  int
	want int
}

fn (err ArgsCountError) msg() string {
	if err.want == 0 {
		return 'Expected no arguments, but got $err.got'
	} else if err.got > err.want {
		return 'Expected at most $err.want arguments, but got $err.got'
	} else {
		return 'Expected at least $err.want arguments, but got $err.got'
	}
}
```

### CODE REVIEW

**Comment by timbasel** [comment]
I don't know if there is a better way to introduce this change more gradually. I haven't found a way to have the `IError` interface accept both the field and the method variant at the same time. But the required change are always really minor

**Comment by timbasel** [comment]
From what I can see all failing tests are just from dependencies that require the PR in their respective repository.
The MSVC failure is because the `make.bat` does not compile itself again after building from `v_win.c` (like in the `Makefile` on Linux) and therefore calling `v.exe self` with the old version. 
This will either fix itself once `v_win.c` is updated or one have add an additional `vc.exe -o v.exe cmd/v` step to the `make.bat`.

**Comment by JalonSolov** [comment]
Can you check to see how this change affects the `coreutils` project?  The `flag` module is used extensively there.

File: vlib/json/json_decode_test.v [hunk 1]
```diff
@@ -14,7 +14,11 @@ mut:
 fn test_json_decode_fails_to_decode_unrecognised_array_of_dicts() {
 	data := '[{"twins":[{"id":123,"seed":"abcde","pubkey":"xyzasd"},{"id":456,"seed":"dfgdfgdfgd","pubkey":"skjldskljh45sdf"}]}]'
 	json.decode(TestTwins, data) or {
-		assert err.msg == "expected field 'twins' is missing"
+		println('testing')
+		dump(err)
+		dump(err.msg())
+		println('test')
```
Comment (spytheman) [inline vlib/json/json_decode_test.v:20]: why are these needed?

Comment (timbasel) [reply vlib/json/json_decode_test.v:20]: forgotten test code, will be removed

**Comment by timbasel** [comment]
Mmh, after having scanned through a bunch of V repositories the dependency on the internals of `IError` (i.e. `.msg` and `.code`) is larger than I initially though. Mainly due to the fact that many have build wrappers around the error to add additional information to the error (this is the exact use case the proposed change handles in a more elegant and general way). But therefore switching everything over at once is not feasible. I will have to think about a better approach tomorrow. If anyone has a an idea how a gradual transition can be accomplished, the help is definitely appreciated.

**Comment by spytheman** [comment]
> If anyone has a an idea how a gradual transition can be accomplished, the help is definitely appreciated.

I can think of "temporary" compiler magic, that will stay for at least a few weeks, that handles `err.msg` by producing code for `err.msg()` *and* producing a notice that it should be changed.

Allso adding a vfmt special case, that will rewrite the old code to the new one, so the process could be automated as much as possible for end users (i.e. in most cases, they should be able to just run `v fmt -w .`).

**Comment by knz** [comment]
I'd love to beta-test this work when it's ready for testing!

**Comment by timbasel** [comment]
You can checkout the commit, the compiler is is building correctly and the whole vlib is updated. The problem I'm currenlty working out is how to make the transition period as seamless as possible.

**Comment by JalonSolov** [comment]
Unless you are a git wizard, I recommend getting familiar with the [Github cli](https://cli.github.com/).  It makes testing PR changes fairly trivial, as it can automatically pull the PR to a branch in your local repo.

**Comment by timbasel** [comment]
With the hack the compiler will produce valid programs regardless if you use the old or new style. The changes of the hack are clearly marked and should be able to be removed easily once the deprecation period ends.

**Review by spytheman — APPROVED** [review]
Excellent work!
